File: code/design-patterns/behavioral/observer/go/observer.go
package main

import "fmt"

// Subscriber interface that all subscribers must implement
type Subscriber interface {
	Update()
}

// Customer is a concrete subscriber that implements the Subscriber interface
type Customer struct {
	name string
}

func (c *Customer) Update() {
	fmt.Printf("%s has been notified.\n", c.name)
}

// Store is the publisher that maintains a list of subscribers and notifies them
type Store struct {
	subscribers []Subscriber
}

func (s *Store) addSubscriber(subscriber Subscriber) {
	s.subscribers = append(s.subscribers, subscriber)
}

func (s *Store) removeSubscriber(subscriber Subscriber) {
	for i, sub := range s.subscribers {
		if sub == subscriber {
			s.subscribers = append(s.subscribers[:i], s.subscribers[i+1:]...)
			break
		}
	}
}

func (s *Store) notify() {
	for _, subscriber := range s.subscribers {
		subscriber.Update()
	}
}

File: code/design-patterns/behavioral/observer/python/observer.py
class Subscriber:
    def update(self):
        raise NotImplementedError("Subclass must implement abstract method")


class Customer(Subscriber):
    def __init__(self, name):
        self.name = name

    def update(self):
        print(f"{self.name} has been notified.")


class Store:
    def __init__(self):
        self.subscribers = []

    def add_subscriber(self, subscriber):
        self.subscribers.append(subscriber)

    def remove_subscriber(self, subscriber):
        self.subscribers.remove(subscriber)

    def notify(self):
        for subscriber in self.subscribers:
            subscriber.update()
